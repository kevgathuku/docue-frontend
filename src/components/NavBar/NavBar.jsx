import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { getSession, initiateLogout } from '../../actions/actionCreators';
import logoSrc from '../../images/favicon.png';

export class NavBar extends React.Component {
  // Receive the current pathname as a prop
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.object,
    loggedIn: PropTypes.bool,
    logoutResult: PropTypes.string,
    pathname: PropTypes.string,
    session: PropTypes.shape({
      loggedIn: PropTypes.bool,
    }),
    token: PropTypes.string,
    user: PropTypes.object,
  };

  componentDidMount() {
    const token = localStorage.getItem('user');
    // Send a request to check if the user is logged in
    this.props.dispatch(getSession(token));

    window.$('.dropdown-button').dropdown();
    window.$('.button-collapse').sideNav();
  }

  componentDidUpdate(prevProps) {
    window.$('.dropdown-button').dropdown();
    // window.$('.button-collapse').sideNav();

    const { logoutResult, session } = this.props;

    if (logoutResult && prevProps.logoutResult !== logoutResult) {
      // Remove the user's token and info
      localStorage.removeItem('user');
      localStorage.removeItem('userInfo');

      this.props.history.push('/');
    }

    if (session && prevProps.session !== session) {
      // session contains 'true' + the user object or 'false'
      if (session.loggedIn === false) {
        // If there is a user token in localStorage, remove it
        // because it is invalid now
        localStorage.removeItem('user');
        localStorage.removeItem('userInfo');

        // If the user is not logged in and is not on the homepage
        // redirect them to the login page
        if (this.props.pathname !== '/') {
          this.props.history.push('/auth');
        }
      } else if (session.loggedIn) {
        // Redirect to the dashboard if logged in and on the auth page or homepage
        if (this.props.pathname === '/auth' || this.props.pathname === '/') {
          this.props.history.push('/dashboard');
        }
      }
    }
  }

  handleLogoutSubmit = (event) => {
    event.preventDefault();
    const token = localStorage.getItem('user');
    this.props.dispatch(initiateLogout(token));
  };

  render() {
    return (
      <nav className="transparent black-text" role="navigation">
        <div className="nav-wrapper container">
          <a className="brand-logo brand-logo-small" href="/">
            <img alt="Docue Logo" id="header-logo" src={logoSrc} />
            {'      Docue'}
          </a>
          <a href="#" data-activates="mobile-demo" className="button-collapse">
            <i className="material-icons" style={{ color: 'grey' }}>
              menu
            </i>
          </a>
          <ul className="side-nav" id="mobile-demo">
            <li>
              <a href="/">Home</a>
            </li>
            <li>
              {this.props.loggedIn ? (
                <a href="/profile">Profile</a>
              ) : (
                <a href="/auth">Login</a>
              )}
            </li>
            <li>
              {this.props.loggedIn ? (
                <a href="/#" onClick={this.handleLogoutSubmit}>
                  Logout
                </a>
              ) : (
                <a href="/auth">Sign Up</a>
              )}
            </li>
          </ul>
          <ul className="right hide-on-med-and-down" id="nav-mobile">
            <li>
              {this.props.loggedIn ? (
                <div>
                  <ul id="dropdown" className="dropdown-content">
                    <li>
                      <a href="/profile">My Profile</a>
                    </li>
                    <li>
                      <a href="/dashboard">All Documents</a>
                    </li>
                    {this.props.user.role &&
                    this.props.user.role.title === 'admin' ? (
                      <li>
                        <a href="/admin">Settings</a>
                      </li>
                    ) : null}
                    <li className="divider" />
                    <li>
                      <a
                        href="/#"
                        id="logout-btn"
                        onClick={this.handleLogoutSubmit}
                      >
                        {' '}
                        Logout
                      </a>
                    </li>
                  </ul>
                  <a
                    className="dropdown-button"
                    data-activates="dropdown"
                    data-beloworigin="true"
                    data-constrainwidth="false"
                  >
                    {this.props.user.name.first}
                    <i className="material-icons right">arrow_drop_down</i>
                  </a>
                </div>
              ) : null}
            </li>
          </ul>
        </div>
      </nav>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    token: state.token,
    user: state.user,
    loggedIn: state.user._id && state.session.loggedIn,
    logoutResult: state.logoutResult,
    session: state.session,
  };
};

export default connect(mapStateToProps)(NavBar);
