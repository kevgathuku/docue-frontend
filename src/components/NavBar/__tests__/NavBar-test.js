import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { NavBar } from '../NavBar.jsx';

describe('NavBar', function() {
  describe('Component Rendering', function() {
    beforeEach(function() {
      global.localStorage.clear();

      window.$ = sinon.stub();
      window.$.withArgs('.dropdown-button').returns(
        sinon.stub({
          dropdown: function() {},
        })
      );
      window.$.withArgs('.button-collapse').returns(
        sinon.stub({
          sideNav: function() {},
        })
      );
    });

    it('renders the correct content', function() {
      const props = {
        dispatch: jest.fn(),
      }
      const wrapper = shallow(<NavBar {...props} />)
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should activate the materialize dropdowns', function(done) {
      const props = {
        dispatch: jest.fn(),
      }
      mount(<NavBar {...props} />);
      // The menu activators should be activated after component mount & update
      expect(window.$.withArgs('.dropdown-button').called).toBe(true);
      expect(window.$.withArgs('.button-collapse').called).toBe(true);
      done();
    });
  });

  describe('Class functions:', function() {
    describe('userSession', function() {
      beforeEach(function() {
        // browserHistory.push = jest.fn();
      });

      it.skip('responds correctly if the user is not logged in', function() {
        sinon.spy(global.localStorage, 'removeItem');

        const props = {
          dispatch: jest.fn(),
          session: {
            loggedIn: false
          }
        }
        shallow(<NavBar {...props} />);

        expect(localStorage.removeItem.withArgs('user').called).toBe(true);
        expect(localStorage.removeItem.withArgs('userInfo').called).toBe(true);
        localStorage.removeItem.restore();
      });

      it.skip('responds correctly if the user is logged in', function() {
        const props = {
          pathname: "/",
          dispatch: jest.fn(),
          history: {
            push: jest.fn(),
          },
          session: {
            loggedIn: true,
            user: {
              name: 'Kevin',
              role: {
                title: 'viewer',
              }
            }
          }
        }
        const wrapper = mount(<NavBar {...props} />);
        // Trigger a change in the UserStore
        // userStore.session = {
        //   loggedIn: 'true',
          // user: {
          //   name: 'Kevin',
          //   role: {
          //     title: 'viewer',
          //   },
          // },
        // };
        // this.props.history.push('/dashboard');
        expect(props.push.mock.calls[0][0]).toBe('/dashboard');
      });
    });
  });

  describe('handleLogout', function() {
    it.skip('should logout the user on click', function() {
      sinon.spy(localStorage, 'removeItem');

      const props = {
        dispatch: jest.fn(),
        user: {
          name: {
            first: 'Kevin',
          },
          role: {
            title: 'viewer',
          },
        },
      }
      mount(<NavBar {...props} />);
      // Trigger a change in the logout store
      // userStore.logoutResult = {
      //   message: 'Successfully logged out',
      // };
      expect(localStorage.removeItem.withArgs('user').called).toBe(true);
      expect(localStorage.removeItem.withArgs('userInfo').called).toBe(true);
    });

    it('should call the logout action on click', function() {
      const mockEvent = {
        preventDefault: jest.fn(),
      };
      const props = {
        loggedIn: true,
        dispatch: jest.fn(),
        token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9',
        user: {
          name: {
            first: 'Kevin',
          },
          role: {
            title: 'viewer',
          },
        },
      };

      let navBar = mount(<NavBar {...props} />);
      // userStore.loginResult = user;
      // The logout button should be in the DOM
      expect(navBar.find('#logout-btn').length).toBe(1);
      navBar.find('#logout-btn').simulate('click', mockEvent);

      expect(mockEvent.preventDefault).toBeCalled();
      expect(props.dispatch).toBeCalled();
    });
  });
});
