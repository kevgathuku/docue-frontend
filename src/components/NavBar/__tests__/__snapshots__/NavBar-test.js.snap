// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Component Rendering renders the correct content 1`] = `
<nav
  className="transparent black-text"
  role="navigation"
>
  <div
    className="nav-wrapper container"
  >
    <a
      className="brand-logo brand-logo-small"
      href="/"
    >
      <img
        alt="Docue Logo"
        id="header-logo"
        src="favicon.png"
      />
            Docue
    </a>
    <a
      className="button-collapse"
      data-activates="mobile-demo"
      href="#"
    >
      <i
        className="material-icons"
        style={
          Object {
            "color": "grey",
          }
        }
      >
        menu
      </i>
    </a>
    <ul
      className="side-nav"
      id="mobile-demo"
    >
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/auth"
        >
          Login
        </a>
      </li>
      <li>
        <a
          href="/auth"
        >
          Sign Up
        </a>
      </li>
    </ul>
    <ul
      className="right hide-on-med-and-down"
      id="nav-mobile"
    >
      <li />
    </ul>
  </div>
</nav>
`;
